Roam: ENG-709: Settings panel for suggestive mode

sid597 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced "Suggestive Mode" and "Suggestive Rules" configuration tabs in the settings interface, providing new options for discourse suggestions.
  * Added interactive panels for managing suggestive rules, including template editing, embedding block references with live previews, and toggling related settings.
  * Enabled management of "page groups" for suggestions, allowing users to add, remove, and organize groups and associated pages.
  * Added controls to include or exclude current page relations and parent/child blocks in suggestion logic.

* **Enhancements**
  * Improved configuration experience with structured panels and real-time UI updates reflecting the latest settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->


https://www.loom.com/share/7218d88262e142098ab0b5622fa25272?sid=94b47b41-2e76-4874-8849-ce78c8cb4baa

sid597 committed:
settings for context and page groups

sid597 committed:
node settings

sid597 committed:
address review

sid597 committed:
fix node creation if it does not exist

sid597 committed:
remove unused, refactor

sid597 committed:
address coderabbit review

sid597 committed:
unify spelling, handle fallback

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,161 @@
+import getSubTree from "roamjs-components/util/getSubTree";
+import { DiscourseNode } from "~/utils/getDiscourseNodes";
+
+const getNodeConfig = (parentUid: string) => {
```
This should be added the `DiscourseNode[]` definition and the parsing shoudl be done in `getDiscourseNodes`

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,161 @@
+  const embeddingRefNode = tree.find((n) =>
+    n.text.startsWith("Embedding Block Ref"),
+  );
+  const match = embeddingRefNode?.children?.[0]?.text?.match(/\(\((.*)\)\)/);
```
Let's support with with or without `((`
EG; `((O08tdFBdx))` or `O08tdFBdx`
You can use roamjs-components `extractRef` function for this.

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,161 @@
+  useEffect(() => {
+    const container = containerRef.current;
+    if (container) {
+      while (container.firstChild) {
```
Is this while/cleanup necessary?

sid597 replied:
no not necessary

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,161 @@
+    if (nodeConfigTree.embeddingRef !== embeddingRef) {
+      setEmbeddingRef(nodeConfigTree.embeddingRef);
+    }
+  }, [nodeConfigTree.embeddingRef]);
```
If we move the getNodeConfig to getDiscourseNodes, then most of this will come with `node`

So the pattern is
- get the exist settings from the db once (in this case, roam blocks)
- when a setting changes
  - update individual setting in state (so you can re-use in this component if needed) (eg: useState)
  - update db (roam blocks)

This pattern then can repeat the next time a user loads this component.

sid597 replied:
Yes, I like it, thanks 👍🏽

mdroidian commented on apps/roam/src/components/settings/Settings.tsx:
```diff
@@ -160,6 +161,12 @@ export const SettingsDialog = ({
+          <Tab
+            id="suggestive-mode-settings"
+            title="Suggestive Mode"
+            className="mb-8 overflow-y-auto"
```
rm `mb-8`, this creates an akward white space

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,161 @@
+
+      <FlagPanel
+        title="First Child"
+        description="If the block is the first child of the template, it will be embedded and ranked."
```
Does this mean the first child of the template?  Or first child of the "Embedding Block Ref"?

If it is the former, why would a user use this instead of just using "Embedding Block Ref" for the first block?

sid597 replied:
yeah first child of "embedding block ref", my bad

mdroidian commented on apps/roam/src/components/settings/NodeConfig.tsx:
```diff
@@ -327,6 +328,23 @@ const NodeConfig = ({
+              <DiscourseNodeSuggestiveRules
+                node={node}
+                parentUid={
+                  getSubTree({
```
Let's move this to the top with the rest.  There's a getUid() pattern that has already been established starting at ln 115

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+    })),
+  );
+
+  const refreshGroups = useCallback(() => {
```
Let's try to follow this pattern:
- get the settings from the db (roam block)
  - in this case that should be getDiscourseNodes
- when a setting changes
  - update and store that individual setting in state (if needed) (eg: useState)
  - update the db (roam blocks)

We shouldn't have to fetch from the db (roam blocks) again.

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+  return (
+    <div className="p-4">
+      <Label>
+        Default Page Groups
```
Let's change this to just "Page Groups" because "Default" suggests it will be the initial state which could then be changed.  But that isn't currently the case here.

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+        Default Page Groups
+        <Description
+          description={
+            "Organize pages into named groups that will be used by default when generating Discourse Suggestions."
```
instead of "used by default" change to "can be selected"

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+            <AutocompleteInput
+              value={newGroupName}
+              setValue={setNewGroupName}
+              placeholder="New group name…"
```
change to "Page Group Name"

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+    autocompleteKeys[groupUid] || 0;
+
+  return (
+    <div className="p-4">
```
rm padding to keep inline with the rest of the elements
<img width="2008" height="928" alt="image" src="https://github.com/user-attachments/assets/1f10b0c6-6b03-4742-bf32-c809c6d8bbfc" />

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,205 @@
+            "Organize pages into named groups that will be used by default when generating Discourse Suggestions."
+          }
+        />
+        <div className="flex flex-col gap-2 pl-2">
```
rm padding to keep inline with the rest of the elements

<img width="2008" height="928" alt="image" src="https://github.com/user-attachments/assets/aba02c85-05f0-4431-99f0-9ab927469894" />

mdroidian commented on apps/roam/src/components/settings/SuggestiveModeSettings.tsx:
```diff
@@ -0,0 +1,70 @@
+          value={settings.suggestiveMode.grabFromReferencedPages.value}
+        />
+        <FlagPanel
+          title="Include Parent And Child Blocks"
```
Does "Include Current Page Relations" make "Include Parent and Child Blocks" redundant?
EG: if "Include Current Page Relations" is checked should "Include Parent and Child Blocks" be checked and disabled?

If so, maybe we make it into a selection instead.  (this can be deferred)

sid597 replied:
Yes this is true : if "Include Current Page Relations" is checked  "Include Parent and Child Blocks" should be checked and disabled.

mdroidian commented on apps/roam/src/components/settings/SuggestiveModeSettings.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  return (
+    <div className="relative flex flex-col gap-4 p-1">
+      <div className="mb-2 text-lg font-semibold text-neutral-dark">
```
RM title.  No other tabs include this so let's stick to the pattern.

mdroidian commented on apps/roam/src/components/settings/SuggestiveModeSettings.tsx:
```diff
@@ -0,0 +1,70 @@
+          text={"Generate & Upload All Node Embeddings"}
+          onClick={() => console.log("Not implemented")}
+          intent={Intent.PRIMARY}
+          style={{ marginTop: "8px" }}
```
use tailwind classes

mdroidian commented on apps/roam/src/components/settings/SuggestiveModeSettings.tsx:
```diff
@@ -0,0 +1,70 @@
+  useEffect(() => {
+    const ensureSettings = async () => {
+      if (!settings.suggestiveMode.parentUid) {
+        await createBlock({
```
Use the pattern mentioned in other comments. (get db settings once)

So here, createBlock will return it's ID.  We can use that to set suggestiveModeUid (or similar) in this components state.  But we don't need to fetch from the db again.  This can result in race conditions and other nonesense we don't want to deal with :)

mdroidian commented on apps/roam/src/utils/getSuggestiveModeConfigSettings.ts:
```diff
@@ -0,0 +1,44 @@
+
+  return {
+    parentUid: suggestiveModeNode.uid,
+    grabFromReferencedPages: getUidAndBooleanSetting({
```
Nit: this name made me think it was a function, as "grab" is close to "get" or "fetch".  Maybe change to something like "includePageRelations".

mdroidian commented on apps/roam/src/utils/getSuggestiveModeConfigSettings.ts:
```diff
@@ -0,0 +1,44 @@
+      tree: suggestiveModeNode.children,
+      text: "Include Current Page Relations",
+    }),
+    grabParentAndChildren: getUidAndBooleanSetting({
```
Nit: this name made me think it was a function, as "grab" is close to "get" or "fetch".  Maybe change to something like "includeParentAndChildBlocks".

mdroidian commented on apps/roam/src/components/settings/NodeConfig.tsx:
```diff
@@ -327,6 +328,23 @@ const NodeConfig = ({
         />
+        <Tab
+          id="node-suggestive-rules"
+          title="Node Suggestive Rules"
```
"Node" is redundant here.
Let's just change it to "Suggestive Mode" to be consistent.

We could add direct links to each node's "Rules" in the Global Setting tab, but that can be deffered.
Likely we'll need more info/docs soon regardless.

mdroidian commented on apps/roam/src/utils/getSuggestiveModeConfigSettings.ts:
```diff
@@ -0,0 +1,44 @@
+      text: "Include Parent And Child Blocks",
+    }),
+    pageGroups: {
+      uid: pageGroupsNode?.uid || "",
```
the `?.uid` and `|| ""` is probably not necessary here.  And it also isn't handled elsewhere so it's misleading.  EG: if `uid` was blank, the code would fail later on.

mdroidian commented on apps/roam/src/utils/getSuggestiveModeConfigSettings.ts:
```diff
@@ -0,0 +1,44 @@
+    }),
+    pageGroups: {
+      uid: pageGroupsNode?.uid || "",
+      name: "Page Groups",
```
Where is this used?  rm if not necessary.

sid597 replied:
it will be used in hyde

mdroidian reviewed (changes requested):
Good start!  Quite a few changes requested but the big one is the settings pattern.  (noted in a few places)

There are comments before this review, so be sure to check those as well (sorry!)

One more big thing, I'm getting a lot of type/build errors.  I tried rm node_modules/package-lock to see if I could get a similar build state as yours, but no luck.  Let's try rebasing this main to update to the latest package-lock.  That means removing your node_modules and using main's package-lock instead of your current one.

Let me know if you have issues.

sid597 committed:
sync

sid597 committed:
address coderabbit code

sid597 committed:
address lint errors

sid597 committed:
use async instead of backend

sid597 committed:
address review

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
address review

sid597 committed:
bug fixes

sid597 committed:
bug fixes

sid597 committed:
Merge branch 'eng-733-glue-ui-components-to-data-extraction-and-database' of https://github.com/DiscourseGraphs/discourse-graph into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
unused import

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' of https://github.com/DiscourseGraphs/discourse-graph into eng-764-bulk-account-uploading

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' of https://github.com/DiscourseGraphs/discourse-graph into eng-764-bulk-account-uploading

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
bulk upsert accounts, use database function imports

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
add comment for future

sid597 committed:
use better import method

sid597 committed:
Merge branch 'eng-733-glue-ui-components-to-data-extraction-and-database' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
commonjs-to-esm

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
fix merge errors, fix tree config setting extraction, test functionality, fix functionality

sid597 committed:
add progress toaster, use space existence condition for on load sync

sid597 committed:
defaultvalue

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,133 @@
+    if (container) {
+      container.innerHTML = "";
+
+      if (uid) {
```
uncessary check

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,133 @@
+          <Button
+            text="Update Embeddings"
+            intent={Intent.NONE}
+            onClick={() => console.log("Not implemented")}
```
👀 Is there a linear ticket or reference for when this will be implemented?

sid597 replied:
🤦🏽 Yeah this needs a ticket to discuss what is the right action for this, I have a minimal function that can be a right approach and was in alpha branch. Thank you for catching this 🙌🏽 


[ENG-800: What is the right approach to handle "Update Embeddings" button?](https://linear.app/discourse-graphs/issue/ENG-800/what-is-the-right-approach-to-handle-update-embeddings-button)

mdroidian commented on apps/roam/src/components/settings/DiscourseNodeSuggestiveRules.tsx:
```diff
@@ -0,0 +1,133 @@
+            text="Update Embeddings"
+            intent={Intent.NONE}
+            onClick={() => console.log("Not implemented")}
+            style={{ minWidth: "140px" }}
```
try to stick with tailwind
protip: if you start any tailwind class, cursor with autocomplete and show you exactly what css will be applied (eg pixel lengths)
<img width="1416" height="397" alt="image" src="https://github.com/user-attachments/assets/8b088e61-d00a-43c1-af40-4b0eecf3cb54" />

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,212 @@
+import getAllPageNames from "roamjs-components/queries/getAllPageNames";
+import { PageGroup } from "~/utils/getSuggestiveModeConfigSettings";
+
+type PageGroupData = PageGroup;
```
Why not just use PageGroup?

sid597 replied:
Yeah I should, maybe a reminiscent of old code, thanks

mdroidian commented on apps/roam/src/components/settings/PageGroupPanel.tsx:
```diff
@@ -0,0 +1,212 @@
+    autocompleteKeys[groupUid] || 0;
+
+  return (
+    <div>
```
Avoid unnessarily in a div if it isn't providing any purpose.  use <> if you just need to wrap it (but in this case you don't need that)

mdroidian commented on apps/roam/src/utils/getDiscourseNodes.ts:
```diff
@@ -5,6 +5,7 @@ import getDiscourseRelations from "./getDiscourseRelations";
 import { roamNodeToCondition } from "./parseQuery";
 import { Condition } from "./types";
 import { InputTextNode, RoamBasicNode } from "roamjs-components/types";
+import extractRef from "roamjs-components/util/extractRef";
```
unused

mdroidian commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -345,9 +370,7 @@ const upsertUsers = async (
+export const createOrUpdateDiscourseEmbedding = async () => {
   const { shouldProceed, lastUpdateTime, worker } = await proposeSyncTask();
 
   if (!shouldProceed) {
```
avoid cluttering up console log

mdroidian commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -111,6 +141,8 @@ export const proposeSyncTask = async (): Promise<SyncTaskInfo> => {
       if (timestamp > now) {
         console.log(
```
avoid cluttering up console log

mdroidian commented on apps/roam/src/index.ts:
```diff
@@ -120,6 +120,8 @@ export default runExtension(async (onloadArgs) => {
   document.addEventListener("input", discourseNodeSearchTriggerListener);
   document.addEventListener("selectionchange", nodeCreationPopoverListener);
 
+  initializeSupabaseSync();
```
Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.eslint@typescript-eslint/no-floating-promises

mdroidian reviewed (approved):
Some thoughts

sid597 committed:
address review, leaving out update embeddings from here

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
update embeddings

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

maparent committed:
ENG-818: Declare dbDotEnv as mjs explicitly

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
Merge branch 'eng-818-make-dbdotenv-explicitly-a-module' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode

sid597 committed:
reverting create a new pr for update embeddings

sid597 committed:
Merge branch 'main' into eng-709-settings-panel-for-suggestive-mode
